yum-repo-manager.py

```
@@ -331,12 +341,172 @@ def manageYumRepo():
+            return False
+        else:
+            ## We know there's possibly a lock and it's expired. We'll take over, providing 
+            ## that the guid of the lock we read as expired is the one we're taking over from. 
+            ## This is an atomic conditional update
+            log("Expired lock found. Attempting to aquire")
+            put_item_params['ExpressionAttributeValues'] = {
+                ':oldguid': {'S': data['Item']['guid']['S']}
+            }
+            put_item_params['ConditionExpression'] = "guid = :oldguid"
+    except Exception as e:
+        log("Exception" + str(e))
+        ## something really bad happened, such as table not found
+        return False
+
+    ## now we're going to try to get the lock. If ANY exception happens, we assume no lock
+    try:
+        db.put_item(**put_item_params)
+        locked = True
+        guid = guid
+        return True
+    except Exception:
+        return False
+
+def releaseDynamoLock(table, lockName):
+    log("Release lock for table [{}] name [{}]".format(table, lockName))
+    global guid, locked
+
+    if not locked:
+        return
+
+    delete_item_params = {
+        'Key': {
+            'name': { 'S': lockName }
+        },
+        'ExpressionAttributeValues': {
+            ':ourguid': {'S': guid}
+        },
+        'TableName': lockTableName,
+        'ConditionExpression': "guid = :ourguid"
+    }
+
+    try:
+        db.delete_item(**delete_item_params)
+        locked = False
+        guid = ""
+    except Exception as e:
+        log(str(e))
+
+def spinlock(table, lockName, locktimeoutMillis, acquireTimeoutMillis):
+    acquireTimeout = millis_in_future(acquireTimeoutMillis)
+    while not acquireDynamoLock(table, lockName, locktimeoutMillis):
+        log("could not acquire lock, sleeping 10 seconds and trying again...")
+        time.sleep(10)
+        if time.time() > acquireTimeout:
+            log("ERROR: acquire lock timeout")
+            sys.exit(1)
```

I don't like sys.exits in the middle of scripts - would rather raise an exception or return False and catch it later